openhands/runtime/utils/edit.py

```diff
@@ -0,0 +1,315 @@
+import os
+import re
+import tempfile
+from abc import ABC, abstractmethod
+
+from openhands.core.config import AppConfig
+from openhands.core.logger import openhands_logger as logger
+from openhands.events.action import (
+    FileEditAction,
+    FileReadAction,
+    FileWriteAction,
+)
+from openhands.events.observation import (
+    ErrorObservation,
+    FileEditObservation,
+    FileReadObservation,
+    FileWriteObservation,
+    Observation,
+)
+from openhands.linter import DefaultLinter
+from openhands.llm.llm import LLM
+from openhands.utils.chunk_localizer import Chunk, get_top_k_chunk_matches
+from openhands.utils.diff import get_diff
+
+SYS_MSG = """Your job is to produce a new version of the file based on the old version and the
+provided draft of the new version. The provided draft may be incomplete (it may skip lines) and/or incorrectly indented. You should try to apply the changes present in the draft to the old version, and output a new version of the file.
+NOTE:
+- The output file should be COMPLETE and CORRECTLY INDENTED. Do not omit any lines, and do not change any lines that are not part of the changes.
+- You should output the new version of the file by wrapping the new version of the file content in a ``` block.
+- If there's no explicit comment to remove the existing code, we should keep them and append the new code to the end of the file.
+"""
+
+USER_MSG = """
+HERE IS THE OLD VERSION OF THE FILE:
+```
+{old_contents}
+```
+
+HERE IS THE DRAFT OF THE NEW VERSION OF THE FILE:
+```
+{draft_changes}
+```
+
+GIVE ME THE NEW VERSION OF THE FILE.
+""".strip()
+
+
+def _extract_code(string):
+    pattern = r'```(?:\w*\n)?(.*?)```'
+    matches = re.findall(pattern, string, re.DOTALL)
+    if not matches:
+        return None
+    return matches[0]
+
+
+def get_new_file_contents(
+    llm: LLM, old_contents: str, draft_changes: str, num_retries: int = 3
+) -> str | None:
+    while num_retries > 0:
+        messages = [
+            {'role': 'system', 'content': SYS_MSG},
+            {
+                'role': 'user',
+                'content': USER_MSG.format(
+                    old_contents=old_contents, draft_changes=draft_changes
+                ),
+            },
+        ]
+        resp = llm.completion(messages=messages)
+        new_contents = _extract_code(resp['choices'][0]['message']['content'])
+        if new_contents is not None:
+            return new_contents
+        num_retries -= 1
+    return None
+
+
+class FileEditRuntimeInterface(ABC):
+    config: AppConfig
+
+    @abstractmethod
+    def read(self, action: FileReadAction) -> Observation:
+        pass
+
+    @abstractmethod
+    def write(self, action: FileWriteAction) -> Observation:
+        pass
+
+
+class FileEditRuntimeMixin(FileEditRuntimeInterface):
+    # Most LLMs have output token limit of 4k tokens.
+    # This restricts the number of lines we can edit to avoid exceeding the token limit.
+    MAX_LINES_TO_EDIT = 300
```

Comment:
Curiously, Sonnet 3.5 has 8k. 🤔 Not in scope of this PR, maybe we can try sometime full file edit when possible, when the file is up to 2-3k...

Reply:
300 lines is roughly around > 2k tokens in my testing & most cheap LLMs can do 4k at least. But qualitatively I find that producing more lines seem to have worse quality. We have some one working on an eval benchmark for this :D - once that's done, we can look at qualitative number and determine the best value for this :)